### DISCUSSION TITLE
Prevent pr creation when dependency update cannot be resolved

### DISCUSSION THREAD

**Discussion by bosix** [renovatebot/renovate#14399] [Request Help] [answered]
Hi there :),

I'm currently facing a question about creating branches and pull requests by the renovate bot.

## Facts

- Renovate Docker Image: renovate/renovate:31.80.1-slim
- Environment: self-hosted Gitlab with self-hosted renovate
- Dependency manager: composer (v2.2)

Renovate-Config:
```json
{
    "extends": [
        "local>xxxxx/_base"
    ],
    "ignorePaths": [
        "**/node_modules/**",
        "**/public/**",
        "**/resources/**",
        "**/tests/**",
        "**/vendor/**"
    ],
    "force": {
        "constraints": {
            "php": "< 8.0.0"
        }
    },
    "packageRules": [
        {
            "matchPackageNames": [
                "laravel/framework",
                "area17/twill",
                "nikic/php-parser"
            ],
            "rangeStrategy": "in-range-only"
        }
    ]
}
```

which extends
```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
        "config:base",
        "group:allNonMajor"
    ],
    "enabled": true,
    "automerge": false,
    "timezone": "Europe/Berlin",
    "branchPrefix": "renovate/",
    "dependencyDashboard": false,
    "prHourlyLimit": 5,
    "prConcurrentLimit": 20,
    "separateMajorMinor": true,
    "separateMultipleMajor": false,
    "labels": [
        "Dependency Update"
    ],
    "updateLockFiles": true,
    "stabilityDays": 1,
    "npm": {
        "stabilityDays": 3
    },
    "docker": {
        "pinDigests": true
    }
}
```

## What renovate does

In one of our projects renovate detects updates for 2 composer dependencies which are grouped into one pull request

![image](https://user-images.githubusercontent.com/26360670/155584542-1a5e9482-7d87-4042-9a3b-6df887babe75.png)

The package `spatie/laravel-permission` has a sub dependency to laravel which is just fine because it's a laravel project. The point is that it is not possible to update this package to v5.5.0 since it collides with another version constraint (the project is pined to laravel <v7 and the update requires at least laravel v7). The renovate bot notices this correctly in a comment inside the pull request:

```
Command failed: docker run --rm --name=renovate_php --label=renovate_child -v "/builds/devops/dependency-watcher/renovate/repos/gitlab/xxxxx":"/builds/devops/dependency-watcher/renovate/repos/gitlab/xxxxx" -v "/builds/devops/dependency-watcher/renovate/cache":"/builds/devops/dependency-watcher/renovate/cache" -e COMPOSER_CACHE_DIR -e COMPOSER_AUTH -w "/builds/devops/dependency-watcher/renovate/repos/gitlab/xxxxx" docker.io/renovate/php:7.4.28 bash -l -c "install-tool composer 2.2.6 && composer update algolia/scout-extended spatie/laravel-permission --with-dependencies --ignore-platform-req='ext-*' --ignore-platform-req='lib-*' --no-ansi --no-interaction --no-scripts --no-autoloader --no-plugins"
Warning: You should avoid overwriting already defined auth settings for gitlab.xxxxx.de.
Loading composer repositories with package information
                                                      Dependency laravel/framework is also a root requirement. Package has not been listed as an update argument, so keeping locked at old version. Use --with-all-dependencies (-W) to include root dependencies.
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Root composer.json requires codeception/module-asserts 2.0.1, found codeception/module-asserts[2.0.1] but the package is fixed to 1.3.1 (lock file version) by a partial update and that version does not match. Make sure you list it as an argument for the update command.
  Problem 2
    - Root composer.json requires spatie/laravel-permission 5.5.0 -> satisfiable by spatie/laravel-permission[5.5.0].
    - spatie/laravel-permission 5.5.0 requires illuminate/auth ^7.0|^8.0|^9.0 -> found illuminate/auth[v7.0.0, ..., 7.x-dev, v8.0.0, ..., 8.x-dev, v9.0.0-beta.1, ..., 9.x-dev] but these were not loaded, likely because it conflicts with another require.

Use the option --with-all-dependencies (-W) to allow upgrades, downgrades and removals for packages currently locked to specific versions.
```

The point is that this pull request keeps open although it cannot be merged until we raise our laravel version (which we will do in future but not yet). Our test pipeline itself is fine because renovate has only updated the compose.json and not the composer.lock which would be necessary for any change in dependencies to take effect.  Only renovates internal ci job `renovate/artifacts` failed because the new dependency could not be resolved.

We could prevent creating the pull request at all by setting `prCreation` to `status-success` but this would also prevent new pull requests for dependency updates which composer can resolve but some tests fail due to e.g. breaking changes.

## What I want renovate to do

I want renovate to check if new dependencies can be resolved by the dependency manager (here: `composer`). If yes, a pull request is created and our devs can have a look at it and even fix tests if some fail due to the change. If dependency-changes cannot be resolved, the pull request should not be created, or even better: the created branch should be deleted or not even created. The best would be if the pull request would only contain valid dependency updates because the first of the two updates mentioned above is 
 kind of blocked because of the second one.

Does anyone have an idea how we can achieve this?

**Comment by rarkins** [comment, marked as answer]
First question: is this incompatibility something which Renovate could potentially have detected using packagist APIs, or is it too deep of a problem that the only way to know is to watch for the error from Composer?

Are you proposing that you want Renovate to suppress PRs anytime "composer install" fails, or just for this specific case?

The downside of this idea is that sometimes you might get into a circular problem where package A's update fails because package B needs to be updated too, and vice versa. With your idea, both PRs would be vanished and you'd be left unaware that you are falling behind with dependencies. I would still recommend that people get the PR, which should show the log message and have a failing status check. That way there's no confusion and little possibility it would get modified by accident. You then have the choice whether to manually update/fix the PR yourself before merging, or to close it to ignore it.

**Reply by bosix** [reply]
In my opinion, it is a thing that Composer usually detects itself when no version is pinned. The newer version is simply not added to `composer.lock`. I am not sure if this is a problem that can be prevented with the packagist api, but it causes a lot of noise because we have 8 prs in this project (even more together with our other projects) that all fail because of almost the same problem. One question in this regard: when we close a pr that says it won't be recreated: is that a kind of "forever" or can we manually trigger the creation again?

![image](https://user-images.githubusercontent.com/26360670/155671793-fad662ac-2729-4bbb-8902-0a174d2a0ad0.png)

I would suggest that there should be a configuration option to decide if prs should be created if composer installation/update fails. Since I think you make a good point, it shouldn't be the default option.

**Reply by rarkins** [reply]
The MR will be created if you rename the existing (closed) one. Also, if you use the Dependency Dashboard, all closed/ignored MRs will be listed there, and you can request a new MR of any of them by clicking a checkbox.

**Reply by bosix** [reply]
Thank you for the advice. I'll have a look at it.

About the option `prCreation` with value `status-success`: the docs say that with that been set a pr will only be created if tests pass. Does this include renovate jobs (`renovate/artifacts`)? The pipeline is marked as "failed" so I assume that a pull request would not be created:
![image](https://user-images.githubusercontent.com/26360670/155674172-8bdbd4da-432f-4a4c-9c30-92c81fc42a31.png)

**Reply by rarkins** [reply]
I think it will include renovate's but I'm not 100% sure

**Reply by bosix** [reply]
Alright. Thank you for your help :)
